### PR TITLE
Initialize selection type in both if/else sections

### LINKED ISSUE #3969: Potential "use before initialization" bug
**Describe the bug**
I ran the valgrind analyser and it showed a "use before initialization" here https://github.com/HDFGroup/hdf5/blob/develop/src/H5Dchunk.c#L3290

The field `fsel_type` was marked as uninitialized by valgrind. The field was created here https://github.com/HDFGroup/hdf5/blob/develop/src/H5Dchunk.c#L1047, after I initialized the field manually the warning disappeared.

Tested on 1.14.3

### PR DESCRIPTION
**Pull request by glennsong09**
The field fsel_type was never initialized in the if section of the if/else initialization code. 

Fixes #3969.

### CODE REVIEW

**Review by derobins — CHANGES_REQUESTED** [review]
I would probably initialize the chunk map in H5D__chunk_io_init() so the struct is fully initialized to sane values in a single location.

You might also change the allocation to use H5FL_CALLOC to avoid future uninitialized fields.

It might also be worth considering initializing the selection type fields to H5S_SEL_ERROR to further avoid uninitialized use.

File: src/H5Dchunk.c [hunk 1]
```diff
@@ -1205,6 +1205,10 @@ H5D__chunk_io_init_selections(H5D_io_info_t *io_info, H5D_dset_io_info_t *dinfo)
         /* Initialize skip list for chunk selections */
         fm->use_single = true;
 
+        /* Initialize selection type in memory and file */
```
Comment (qkoziol) [inline src/H5Dchunk.c]: I suggest that you put this in H5D__chunk_io_init() instead, so that it doesn't get missed for another case in the future.

**Review by qkoziol — CHANGES_REQUESTED** [review]

File: src/H5Dchunk.c (hunk 1 above)
Comment (derobins) [reply src/H5Dchunk.c]: I see we are both working this evening :)

**Comment by qkoziol** [comment]
LOL - @derobins and I with the same comment, simultaneously

**Comment by jhendersonHDF** [comment]
> You might also change the allocation to use H5FL_CALLOC to avoid future uninitialized fields.

I believe the chunk map used to be allocated with calloc and it was switched back to malloc because the allocation overhead of calloc in a tight loop from a Mathworks example was proving to be a bit much.

**Comment by derobins** [comment]
> > You might also change the allocation to use H5FL_CALLOC to avoid future uninitialized fields.
> 
> I believe the chunk map used to be allocated with calloc and it was switched back to malloc because the allocation overhead of calloc in a tight loop from a Mathworks example was proving to be a bit much.

That's fine. Just setting the two values in that function would be a good fix.

**Comment by glennsong09** [comment]
> > > You might also change the allocation to use H5FL_CALLOC to avoid future uninitialized fields.
> > 
> > 
> > I believe the chunk map used to be allocated with calloc and it was switched back to malloc because the allocation overhead of calloc in a tight loop from a Mathworks example was proving to be a bit much.
> 
> That's fine. Just setting the two values in that function would be a good fix.

Okay, I updated the initialization location and values.

**Review by jhendersonHDF — APPROVED** [review]

**Review by derobins — APPROVED** [review]

**Review by qkoziol — APPROVED** [review]

**Pull request merged**